datastore/leveldb/leveldbds.nim

```diff
@@ -0,0 +1,149 @@
+import std/options
+import std/tables
+import std/os
+import std/strformat
+
+import pkg/leveldbstatic
+import pkg/chronos
+import pkg/questionable
+import pkg/questionable/results
+import pkg/stew/byteutils
+from pkg/stew/results as stewResults import isErr
+import pkg/upraises
+
+import ../datastore
+import ../defaultimpl
+
+push: {.upraises: [].}
+
+type
+  LevelDbDatastore* = ref object of Datastore
+    db: LevelDb
+    locks: TableRef[Key, AsyncLock]
+
+method has*(self: LevelDbDatastore, key: Key): Future[?!bool] {.async.} =
+  try:
+    let str = self.db.get($key)
+    return success(str.isSome)
+  except LevelDbException as e:
+    return failure("LevelDbDatastore.has exception: " & e.msg)
+
+method delete*(self: LevelDbDatastore, key: Key): Future[?!void] {.async.} =
+  try:
+    self.db.delete($key, sync = true)
+    return success()
+  except LevelDbException as e:
+    return failure("LevelDbDatastore.delete exception: " & e.msg)
+
+method delete*(self: LevelDbDatastore, keys: seq[Key]): Future[?!void] {.async.} =
+  for key in keys:
+    if err =? (await self.delete(key)).errorOption:
+      return failure(err.msg)
+  return success()
+
+method get*(self: LevelDbDatastore, key: Key): Future[?!seq[byte]] {.async.} =
+  try:
+    let str = self.db.get($key)
+    if not str.isSome:
+      return failure(newException(DatastoreKeyNotFound, "LevelDbDatastore.get: key not found " & $key))
+    let bytes = str.get().toBytes()
+    return success(bytes)
+  except LevelDbException as e:
+    return failure("LevelDbDatastore.get exception: " & $e.msg)
+
+method put*(self: LevelDbDatastore, key: Key, data: seq[byte]): Future[?!void] {.async.} =
+  try:
+    let str = string.fromBytes(data)
+    self.db.put($key, str)
+    return success()
+  except LevelDbException as e:
+    return failure("LevelDbDatastore.put exception: " & $e.msg)
+
+method put*(self: LevelDbDatastore, batch: seq[BatchEntry]): Future[?!void] {.async.} =
+  for entry in batch:
+    if err =? (await self.put(entry.key, entry.data)).errorOption:
```

Comment:
This should be using the batching operations from leveldb.

Reply:
Good catch. This was left-over from my initial hacky attempt to plug leveldb into the datastore interface.
Your comments are implemented here: https://github.com/codex-storage/nim-datastore/pull/65/files